### ISSUE TITLE
Add Reference Tag support for Campaign elements

### ISSUE THREAD

**Issue by andrewmenich** [putyourlightson/craft-campaign#218]
It would be nice to have Reference Tag support for the custom elements created by Campaign (campaign, sendout, contact,  segment). The need for this came up as I was working through creating a rich-text field for an email campaign that would allow our client to add inline contact variables, e.g.

![{contact:{*contactId*}:firstName}﻿ ](https://cdn2.andrewmeni.ch/eyJidWNrZXQiOiJhbmRyZXdtZW5pLmNoIiwia2V5IjoiU2NyZWVuLVNob3QtMjAyMS0wMi0yNC1hdC01LjIzLjMyLVBNLnBuZyJ9)

The inspiration came from Andrew Welch's Rich Variables plugin, which uses reference tags under the hood. I would be happy to submit a PR for this change. It looks like adding support for them is [fairly trivial](https://craftcms.com/docs/3.x/extend/element-types.html#reference-tags). 

Thanks as always for a great plugin.

**Comment by bencroker** [comment]
I'd be happy to add the reference tags, but not sure they will be available in the Rich Variables plugin, which appears to only work with globals.

> Rich Variables allows you to easily use Craft Globals as variables in Rich Text fields.

https://nystudio107.com/docs/rich-variables/#rich-variables-overview

**Comment by bencroker** [comment]
Wouldn't you want to use the `contact` variable that is already available in campaigns?

```twig
Hello {{ contact.firstName }},
```

**Comment by andrewmenich** [comment]
Thanks for your consideration. Just to clarify - I'm not using that plugin, but rather my own Redactor plugin that is modeled after that one. We have set up a content builder matrix for our client with different blocks they can use to compose an email. Basically one campaign type and template to rule them all. One of these blocks is a Redactor field with a Variables plugin that adds a handful of reference tags inline with the rest of the copy.  I'm already using this method with great success for some native Craft elements that are related to a given contact, but now we'd like to be able to add some variables for a Contact element.

Example:
![email block builder](https://cdn2.andrewmeni.ch/eyJidWNrZXQiOiJhbmRyZXdtZW5pLmNoIiwia2V5IjoiU2NyZWVuLVNob3QtMjAyMS0wMi0yNS1hdC04LjU3LjA5LUFNLnBuZyJ9)

Preview:
![email preview](https://cdn2.andrewmeni.ch/eyJidWNrZXQiOiJhbmRyZXdtZW5pLmNoIiwia2V5IjoiU2NyZWVuLVNob3QtMjAyMS0wMi0yNS1hdC04LjU3LjM2LUFNLnBuZyJ9)

MJML: (userId, locationId, etc. are set previously in the email template)
```
<mj-section padding="0">
    <mj-column>
        <mj-text padding="30px 60px 15px" mj-class="sentinel">
            {% set modifiedRichText = block.copy|replace({'{*userId*}': userId, '{*locationId*}': locationId) %}
            {{modifiedRichText|parseRefs|raw}}
        </mj-text>
    </mj-column>
</mj-section>
```

Hopefully this clarifies things. I know it's an edge case but thought I'd ask!

**Comment by bencroker** [comment]
Makes sense, thanks for clarifying! Added in https://github.com/putyourlightson/craft-campaign/commit/c2b955ba83fd1ae096b025b81d34d92026ad4b36 and released in version 1.19.1.
